File: zhihu/zhihu-backend/app/schemas/contract.py
from pydantic import BaseModel, Field
from typing import Optional, List, Any
from datetime import datetime


class ContractCreate(BaseModel):
    case_id: Optional[int] = None
    career_event_id: Optional[int] = None
    source_action_id: Optional[int] = None
    linked_offer_id: Optional[int] = None
    source_attachment_id: Optional[int] = None
    display_name: Optional[str] = None
    document_kind: str = "labor_contract"
    status: str = "active"
    parse_status: str = "ready"
    parse_mode: Optional[str] = None
    parse_notice: Optional[str] = None
    parse_error_code: Optional[str] = None
    page_count: Optional[int] = None
    text_page_count: Optional[int] = None
    ocr_page_count: Optional[int] = None
    parse_quality: dict[str, Any] = Field(default_factory=dict)
    employer: Optional[str] = None
    contract_term: Optional[str] = None
    probation: Optional[str] = None
    salary_terms: Optional[str] = None
    work_location: Optional[str] = None
    working_hours: Optional[str] = None
    non_compete: Optional[str] = None
    penalty_terms: Optional[str] = None
    termination_terms: Optional[str] = None
    raw_text: Optional[str] = None


class ContractResponse(BaseModel):
    id: int
    case_id: Optional[int] = None
    career_event_id: Optional[int] = None
    linked_offer_id: Optional[int] = None
    source_attachment_id: Optional[int] = None
    display_name: Optional[str] = None
    document_kind: str = "labor_contract"
    status: str = "active"
    parse_status: str = "ready"
    parse_mode: Optional[str] = None
    parse_notice: Optional[str] = None
    parse_error_code: Optional[str] = None
    page_count: Optional[int] = None
    text_page_count: Optional[int] = None
    ocr_page_count: Optional[int] = None
    parse_quality: dict[str, Any] = Field(default_factory=dict)
    employer: Optional[str] = None
    contract_term: Optional[str] = None
    probation: Optional[str] = None
    salary_terms: Optional[str] = None
    work_location: Optional[str] = None
    working_hours: Optional[str] = None
    non_compete: Optional[str] = None
    penalty_terms: Optional[str] = None
    termination_terms: Optional[str] = None
    raw_text: Optional[str] = None
    archived_at: Optional[datetime] = None
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None

    class Config:
        from_attributes = True


class ContractPasteCreate(BaseModel):
    text: str
    display_name: Optional[str] = None
    document_kind: str = "auto"
    linked_offer_id: Optional[int] = None
    career_event_id: Optional[int] = None
    source_action_id: Optional[int] = None
    auto_review: bool = True


class ContractUpdate(BaseModel):
    display_name: Optional[str] = None
    document_kind: Optional[str] = None
    linked_offer_id: Optional[int] = None
    status: Optional[str] = None


class ContractReviewSnapshotResponse(BaseModel):
    id: int
    contract_id: int
    attachment_version_id: Optional[int] = None
    review_number: int
    extracted_fields: dict[str, Any]
    findings: List[Any]
    summary: str
    review_mode: str
    rule_version: str
    clause_segments: List[Any] = Field(default_factory=list)
    provider_name: Optional[str] = None
    model_name: Optional[str] = None
    prompt_version: Optional[str] = None
    redaction_version: Optional[str] = None
    ai_status: str = "not_requested"
    ai_input_clause_count: int = 0
    ai_batch_count: int = 0
    ai_completed_batch_count: int = 0
    redaction_report: dict[str, Any] = Field(default_factory=dict)
    coverage_report: dict[str, Any] = Field(default_factory=dict)
    created_at: datetime

    class Config:
        from_attributes = True


class ContractLinkedOfferSummary(BaseModel):
    id: int
    name: Optional[str] = None
    company_name: Optional[str] = None
    job_title: Optional[str] = None


class ContractDetailResponse(ContractResponse):
    latest_review: Optional[ContractReviewSnapshotResponse] = None
    review_count: int = 0
    linked_offer: Optional[ContractLinkedOfferSummary] = None
    linked_offer_contract_count: int = 0
    linked_offer_contract_index: Optional[int] = None


class ContractReviewResponse(BaseModel):
    contract_id: int
    snapshot_id: int
    review_number: int
    findings: List[Any]
    extracted_fields: dict[str, Any]
    summary: str
    important_count: int
    review_count: int
    reused: bool = False
    reviewed_at: datetime
    synced_finding_count: int = 0
    synced_action_count: int = 0


class ContractFollowUpMessage(BaseModel):
    role: str
    content: str = Field(min_length=1, max_length=800)


class ContractFollowUpRequest(BaseModel):
    clause_id: str = Field(min_length=1, max_length=100)
    finding_code: str = Field(min_length=1, max_length=100)
    question: str = Field(min_length=2, max_length=600)
    history: List[ContractFollowUpMessage] = Field(default_factory=list, max_length=6)


class ContractFollowUpResponse(BaseModel):
    answer: str
    evidence_quote: Optional[str] = None
    limits: str
    provider_name: str
    model_name: str
    prompt_version: str
    redaction_version: str
    review_method: str


class ContractFollowUpHistoryItem(BaseModel):
    id: int
    turn_number: int
    question: str
    answer: str
    evidence_quote: Optional[str] = None
    limits: str
    provider_name: Optional[str] = None
    model_name: Optional[str] = None
    prompt_version: Optional[str] = None
    redaction_version: Optional[str] = None
    review_method: Optional[str] = None
    created_at: datetime

    class Config:
        from_attributes = True


class ContractFollowUpHistoryResponse(BaseModel):
    contract_id: int
    review_snapshot_id: int
    clause_id: str
    finding_code: str
    items: List[ContractFollowUpHistoryItem] = Field(default_factory=list)


class ConsistencyDiff(BaseModel):
    field: str
    offer_value: Optional[str] = None
    contract_value: Optional[str] = None
    status: str
    note: Optional[str] = None


class ConsistencyResponse(BaseModel):
    contract_id: int
    offer_id: int
    diffs: List[Any]
    consistent_count: int
    issue_count: int
    synced_finding_count: int = 0
    synced_action_count: int = 0
    review_mode: str = "rules_only"
    model_status: str = "not_requested"
    provider_name: Optional[str] = None
    model_name: Optional[str] = None
    prompt_version: Optional[str] = None
    redaction_version: Optional[str] = None


class ChecklistItem(BaseModel):
    priority: str
    title: str
    reason: str
    script: Optional[str] = None


class ChecklistResponse(BaseModel):
    contract_id: int
    checklist: List[Any]
    synced_action_count: int = 0

File: zhihu/zhihu-frontend/src/types/contract.ts
export type ContractAttention = "important" | "review" | "note";

export interface ContractEvidence {
  text: string;
  start: number | null;
  end: number | null;
  excerpt_start?: number | null;
  excerpt_end?: number | null;
}

export interface ContractClauseSegment {
  id: string;
  order: number;
  title: string;
  category: string;
  text: string;
  start: number;
  end: number;
  page_start?: number | null;
  page_end?: number | null;
}

export interface ContractFinding {
  code: string;
  clause_id?: string | null;
  category: string;
  title: string;
  attention: ContractAttention;
  explanation: string;
  next_step: string;
  evidence: ContractEvidence;
  source: string;
  confidence?: number;
  redacted_evidence_quote?: string;
  rule_code?: string;
}

export interface ExtractedContractField {
  label: string;
  value: string | null;
  status: "extracted" | "candidate" | "blank_in_source" | "unknown";
  source: ContractEvidence | null;
  quality_note?: string | null;
}

export interface ContractReviewSnapshot {
  id: number;
  contract_id: number;
  attachment_version_id: number | null;
  review_number: number;
  extracted_fields: Record<string, ExtractedContractField>;
  findings: ContractFinding[];
  summary: string;
  review_mode: string;
  rule_version: string;
  clause_segments: ContractClauseSegment[];
  provider_name: string | null;
  model_name: string | null;
  prompt_version: string | null;
  redaction_version: string | null;
  ai_status: "queued" | "running" | "success" | "partial_success" | "unavailable" | "failed" | "privacy_blocked" | "no_relevant_clauses" | "not_requested";
  ai_input_clause_count: number;
  ai_batch_count: number;
  ai_completed_batch_count: number;
  redaction_report: Record<string, unknown>;
  coverage_report: Record<string, unknown>;
  created_at: string;
}

export interface ContractRecord {
  id: number;
  case_id: number;
  career_event_id: number | null;
  linked_offer_id: number | null;
  source_attachment_id: number | null;
  display_name: string | null;
  document_kind: string;
  status: "active" | "archived";
  parse_status: "extracting" | "processing" | "reviewing" | "ready" | "failed";
  parse_mode: string | null;
  parse_notice: string | null;
  parse_error_code: string | null;
  page_count: number | null;
  text_page_count: number | null;
  ocr_page_count: number | null;
  parse_quality: Record<string, unknown> & {
    document_profile?: "labor_contract" | "special_agreement" | "employee_handbook" | "other_employment_document";
    document_kind_detection?: {
      status?: "detected" | "needs_confirmation" | "manual";
      value?: string | null;
      source?: "local_text";
      was_automatic?: boolean;
    };
  };
  employer: string | null;
  contract_term: string | null;
  probation: string | null;
  salary_terms: string | null;
  work_location: string | null;
  working_hours: string | null;
  non_compete: string | null;
  penalty_terms: string | null;
  termination_terms: string | null;
  raw_text: string | null;
  archived_at: string | null;
  created_at: string | null;
  updated_at: string | null;
  latest_review: ContractReviewSnapshot | null;
  review_count: number;
  linked_offer: {
    id: number;
    name: string | null;
    company_name: string | null;
    job_title: string | null;
  } | null;
  linked_offer_contract_count: number;
  linked_offer_contract_index: number | null;
}

export interface ContractReviewResponse {
  contract_id: number;
  snapshot_id: number;
  review_number: number;
  findings: ContractFinding[];
  extracted_fields: Record<string, ExtractedContractField>;
  summary: string;
  important_count: number;
  review_count: number;
  reused: boolean;
  reviewed_at: string;
  synced_finding_count: number;
  synced_action_count: number;
}
